client/a8c-for-agencies/sections/referrals/lib/get-estimated-commission.ts

```diff
@@ -0,0 +1,86 @@
+import { APIProductFamilyProduct } from '../../../../state/partner-portal/types';
+import { Referral } from '../types';
+import { getNextPayoutDateActivityWindow } from './get-next-payout-date';
+
+export const getProductCommissionPercentage = ( slug?: string ) => {
+	if ( ! slug ) {
+		return 0;
+	}
+	if ( [ 'wpcom-hosting', 'pressable-hosting' ].includes( slug ) ) {
+		return 0.2;
+	}
+	if ( slug.startsWith( 'jetpack-' ) || slug.startsWith( 'woocommerce-' ) ) {
+		return 0.5;
+	}
+	return 0;
+};
+
+export const getDailyPrice = ( product: APIProductFamilyProduct, quantity: number ) => {
+	// If quantity is not 1 than we search corresponding bundle
+	if ( quantity !== 1 ) {
+		return (
+			product.supported_bundles.find( ( bundleProduct ) => quantity === bundleProduct.quantity )
+				?.price_per_unit ?? 0
+		);
+	}
+	// If quantity is 1 than we return price per unit
+	return product.price_per_unit ?? 0;
+};
+
+export const getEstimatedCommission = (
+	referrals: Referral[],
+	products: APIProductFamilyProduct[],
+	date: Date = new Date()
+) => {
+	const activityWindow = getNextPayoutDateActivityWindow( date );
+
+	const totalCommissionInCents = referrals.reduce( ( acc, referral ) => {
+		if ( ! referral?.purchases?.length ) {
+			return acc;
+		}
+		for ( const purchase of referral.purchases ) {
+			// We go over 'active' and 'cancelled' subscriptions
+			if ( ! purchase || purchase.status === 'pending' ) {
+				continue;
+			}
+			// Find corresponding product for price
+			const product = products.find( ( product ) => purchase.product_id === product.product_id );
+			if ( ! product ) {
+				continue;
+			}
+			// Day the license was issued
+			const issuedDate = new Date( purchase.license.issued_at );
+			// Day the license was revoked if present
+			const revokedAt = purchase.license.revoked_at
+				? new Date( purchase.license.revoked_at )
+				: null;
+
+			// Start date is the latest of the license issued date and activity window start
+			const start = Math.max( issuedDate.getTime(), activityWindow.start.getTime() );
+			// Finish date is the earliest of the license revoked date and activity window finish
+			const finish = Math.min(
+				revokedAt ? revokedAt.getTime() : Infinity,
+				activityWindow.finish.getTime()
+			);
+			// Total days is the difference between finish and start dates in days
+			const totalDays = Math.floor( ( finish - start ) / ( 1000 * 60 * 60 * 24 ) );
+			if ( totalDays < 1 ) {
+				continue;
+			}
+
+			const dailyPrice = getDailyPrice( product, purchase.quantity );
```

Comment:
Looking at `product` and `purchase.quantity`. If I understand this right, a referral can have multiple purchases. And each purchase corresponds to a single product, where quantity equals the number of that product in the given purchase. Is that right? In other words, a purchase always relates to a single product rather than multiple different products.

Reply:
Yes, that's correct. Basically, we probably can think of `purchase` as a license. 
Or, a referral shopping cart, each product inside shopping cart will become it's own purchase.
It's also easy to see in dev console for `/referrals` endpoint.